pkg/controller/ipam/antrea_ipam_controller.go

```diff
@@ -152,31 +161,67 @@ func (c *AntreaIPAMController) enqueueStatefulSetDeleteEvent(obj interface{}) {
 // Inspect all IPPools for stale IP Address entries.
 // This may happen if controller was down during StatefulSet delete event.
 // If such entry is found, enqueue cleanup event for this StatefulSet.
-func (c *AntreaIPAMController) cleanupStaleStatefulSets() {
+func (c *AntreaIPAMController) cleanupStaleAddresses() {
 	pools, _ := c.ipPoolLister.List(labels.Everything())
 	lister := c.statefulSetInformer.Lister()
 	statefulSets, _ := lister.List(labels.Everything())
 	statefulSetMap := make(map[string]bool)
 
+	klog.InfoS("Cleanup job for IP Pools started")
+
 	for _, ss := range statefulSets {
 		// Prepare map of existing StatefulSets for quick reference below
 		statefulSetMap[k8s.NamespacedName(ss.Namespace, ss.Name)] = true
 	}
 
+	poolsUpdated := 0
 	for _, ipPool := range pools {
+		updateNeeded := false
+		var newList []crdv1a2.IPAddressState
 		for _, address := range ipPool.Status.IPAddresses {
+			// Cleanup reserved addresses
+			addressCopy := address.DeepCopy()
+			if address.Owner.Pod != nil {
+				_, err := c.podLister.Pods(address.Owner.Pod.Namespace).Get(address.Owner.Pod.Name)
+				if err != nil && errors.IsNotFound(err) {
+					klog.InfoS("IPPool contains stale IPAddress for Pod that no longer exists", "IPPool", ipPool.Name, "Namespace", address.Owner.Pod.Namespace, "Pod", address.Owner.Pod.Name)
+					addressCopy.Owner.Pod = nil
+					if addressCopy.Owner.StatefulSet != nil {
+						addressCopy.Phase = crdv1a2.IPAddressPhaseReserved
+					}
+					updateNeeded = true
+				}
+			}
 			if address.Owner.StatefulSet != nil {
 				key := k8s.NamespacedName(address.Owner.StatefulSet.Namespace, address.Owner.StatefulSet.Name)
 				if _, ok := statefulSetMap[key]; !ok {
 					// This entry refers to StatefulSet that no longer exists
-					klog.InfoS("IPPool contains stale IPAddress for StatefulSet that no longer exists", "IPPool", ipPool.Name, "StatefulSet", key)
-					c.statefulSetQueue.Add(key)
-					// Mark this entry in map to ensure cleanup is enqueued only once
-					statefulSetMap[key] = true
+					klog.InfoS("IPPool contains stale IPAddress for StatefulSet that no longer exists", "IPPool", ipPool.Name, "Namespace", address.Owner.StatefulSet.Namespace, "StatefulSet", address.Owner.StatefulSet.Name)
+					addressCopy.Owner.StatefulSet = nil
+					updateNeeded = true
+
 				}
 			}
+
+			if addressCopy.Owner.StatefulSet != nil || addressCopy.Owner.Pod != nil {
+				newList = append(newList, *addressCopy)
+			}
+		}
+
+		if updateNeeded {
+			ipPool.Status.IPAddresses = newList
```

Comment:
The race is in the code that waits for the Pool to clean up.. I added sleep to avoid it. Not sure if there is a more robust way without adding complexity to the code.

Reply:
Objects returned by Lister must be treated as read-only, otherwise there would by race condition as long as there are more than 1 consumers for the Lister.
https://github.com/antrea-io/antrea/blob/91015bb7ee7597eac287bc28adbd19f6c62b6781/pkg/client/listers/crd/v1alpha2/ippool.go#L30

The code should make a copy for the entire pool before updating any of its field. And copying at L183 can be removed.

Reply:
Got it, thanks @tnqn